### PR TITLE
Response from /send_join now use gomatrixserverlib.RespSendJoin

### LINKED ISSUE #614: Make response from /send_join use gomatrixserverlib.RespSendJoin

### PR DESCRIPTION
**Pull request by behouba**
Signed-off-by: Kouame Behouba Manassé behouba@gmail.com

Should fix #614

### Pull Request Checklist

* [ ] I have added any new tests that need to pass to `testfile` as specified in [docs/sytest.md](https://github.com/matrix-org/dendrite/blob/master/docs/sytest.md)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/dendrite/blob/master/CONTRIBUTING.md#sign-off)

### CODE REVIEW

**Comment by Cnly** [comment]
This branch is 49 commits behind `matrix-org:master`. Could you do a merge?

**Comment by anoadragon453** [comment]
@Cnly have merged forward and ran CI

**Comment by Cnly** [comment]
Maybe gomatrixserverlib needs some changes? It seems [`RespSendJoin`](https://github.com/matrix-org/gomatrixserverlib/blob/HEAD/federationtypes.go#L330) is missing the [required](https://matrix.org/docs/spec/server_server/r0.1.2#put-matrix-federation-v1-send-join-roomid-eventid) field `origin`.

**Comment by behouba** [comment]
@Cnly you are right, after reading the [spec](https://matrix.org/docs/spec/server_server/r0.1.3#put-matrix-federation-v1-send-join-roomid-eventid) there is a missing `origin` field.

**Comment by anoadragon453** [comment]
@behouba Would you be up for writing a PR for gomatrixserverlib to add the field?

**Comment by behouba** [comment]
I will work on it with pleasure :)

**Comment by behouba** [comment]
@anoadragon453 Should i close this PR?

**Comment by anoadragon453** [comment]
> @anoadragon453 Should i close this PR?

Nah, once gomatrixserverlib is updated you can change this PR to set the fields as well as update Dendrite's gomatrixserverlib's version :)

**Comment by behouba** [comment]
Ok, great ! 👍

**Comment by behouba** [comment]
In the later commit i have updated gomatrixserverlib version and i have also updated some parts of syncapi for compatibility with the new gomatrixserverlib. hoping i did it properly
@Cnly @anoadragon453

File: federationapi/routing/join.go [hunk 1]
```diff
@@ -176,9 +176,10 @@ func SendJoin(
 
 	return util.JSONResponse{
 		Code: http.StatusOK,
-		JSON: map[string]interface{}{
-			"state":      stateAndAuthChainRepsonse.StateEvents,
-			"auth_chain": stateAndAuthChainRepsonse.AuthChainEvents,
+		JSON: gomatrixserverlib.RespSendJoin{
+			StateEvents: stateAndAuthChainRepsonse.StateEvents,
```
Comment (abhishekkumar2718) [inline federationapi/routing/join.go]: @babolivier 
Please correct the spelling of `stateAndAuthChainRepsonse` to `stateAndAuthChainResponse` in this function.

**Comment by Cnly** [comment]
I did a merge from master onto this branch to resolve the current conflicts and keep the PR ready. Some more changes may still be needed after the CI stuff is fixed though.

**Comment by Kegsay** [comment]
@neilalexander as part of trying to clear the PR list, can I pin this on you to look at next week please given it touches a lot of federation bits you have worked on recently?

**Pull request merged**